### ISSUE TITLE
Incorrect metrics and/or interpolation when rendering variable fonts

### ISSUE THREAD

**Issue by rschiang** [lapce/lapce#1175]
## Lapce Version

0.2.0

## System information

macOS Monterey 12.5.1

## Describe the bug

<img width="502" alt="Source Sans 3 VF" src="https://user-images.githubusercontent.com/1512231/189711455-331fa65f-cb6a-4218-9647-5be717812211.png"><br><small><em>(A problematic rendering of Source Sans 3 VF, illustrating the following issues)</em></small>

### 1. Font weights aren’t correctly rendered when picking a variable font as the UI font.

<details>
<summary>e.g., SF Pro, Inter</summary>
<img width="491" alt="SF Pro" src="https://user-images.githubusercontent.com/1512231/189706521-05ce87a2-c78c-4af1-a713-70bb3e0a15a6.png">
<img width="518" alt="Inter" src="https://user-images.githubusercontent.com/1512231/189706206-aeda2994-7367-4c9d-b1c1-6efb29e2e6f7.png">
</details>

<details>
<summary>c.f. Regular, standalone fonts (Ubuntu, Helvetica Neue)</summary>
<img width="498" alt="Ubuntu" src="https://user-images.githubusercontent.com/1512231/189706799-48ae4a89-50a5-4b29-b8f3-853368c18f48.png">
<img width="508" alt="Helvetica Neue" src="https://user-images.githubusercontent.com/1512231/189706141-168b5ef6-b9c9-4755-9063-aeca69fe0e86.png">
</details>

### 2. Font interpolation breaks under specific variable fonts.

<details>
<summary>e.g., Source Code Variable, Bitter</summary>
<img width="509" alt="Source Code Variable" src="https://user-images.githubusercontent.com/1512231/189706352-debb4d65-f72a-4d29-a0dc-86ea886fbba3.png">
<img width="534" alt="Bitter" src="https://user-images.githubusercontent.com/1512231/189708435-924c5267-32ae-4b76-a738-ffe1a358ab81.png">
</details>

## Additional information

All fonts mentioned in this bug report are OFL-licensed fonts except for SF Pro and Helvetica Neue.

Although Apple fonts (SF Pro) is involved in this bug report, this problem is different from #575 (incorrect metrics with Fira Code and SF Mono), as each weight of SF Mono is distributed as its own OTF file.

<details>
<summary>What SF Mono would look like</summary>
<img width="521" alt="SF Mono" src="https://user-images.githubusercontent.com/1512231/189711149-42eb3e40-d3e2-402d-8e18-0dd783ab5105.png">
</details>

This might be an upstream font engine issue, but I’m really new to Rust UI libraries, so I’ll start the discussion here and see where exactly we could redirect the bug report to.

**Comment by rschiang** [comment]
Seems like the upstream Piet library (which Druid UI uses) already has variable font support (linebender/piet#258), but Druid UI might have some issues on picking the correct weight (linebender/druid#1573).

**Comment by Okirshen** [comment]
![image](https://user-images.githubusercontent.com/44813093/189736198-f65e6f81-0fc8-46c7-83bc-53787f52db80.png)
Similar results with FiraCode Nerd Font

**Comment by rschiang** [comment]
I’ve run against linebender/piet@45ddf3a8229c5384c3119a7c13b8ceb9b124cefb and linebender/druid@79eb7749c96550fe16e4ef36e212f4275af3977e (both the current `HEAD`) and **none** of the problem occurred. 

The problem seems to reside in the custom Druid fork (and the `lapce/piet-wgpu` font rendering implementation), as shown in the follows:

##### Variable font support on Piet `HEAD`

<img width="240" alt="Piet v0.5.0 coregraphics-test-14" src="https://user-images.githubusercontent.com/1512231/190225888-f984bb79-f070-4c54-b6cd-95d97e25a2cf.png"><img width="240" alt="Piet v0.5.0 cairo-test-14-" src="https://user-images.githubusercontent.com/1512231/190226129-287b2ae9-41f3-4eae-ad3f-4698ac061f3a.png">

##### Variable font support on Druid `HEAD`

<img width="500" alt="Druid v0.7.0 custom_widget Example" src="https://user-images.githubusercontent.com/1512231/190232909-65c0e56e-6e14-41b4-bea3-fd09b03ed6b6.png">

##### Variable font support on lapce/piet-wgpu@94a8f01262d59ac2546111beb4a57cbdd654f3d6 custom fork

<img width="240" alt="Piet v0.4.0 wgpu-test-14" src="https://user-images.githubusercontent.com/1512231/190246828-dd64f184-2f23-4a9a-a152-8389fd42f036.png">

##### Variable font support on lapce/druid@ae9a85d2430a53255ae05c20eac8b0241924f293 custom fork

<img width="500" alt="lapce/druid shell_opengl branch custom_widget Example (image removed due to the lack of backend support)" src="https://user-images.githubusercontent.com/1512231/190237826-25a59d27-c3cc-4775-856c-114647bc8fe1.png">

**TL;DR**: It’s a rendering backend problem. 
(CC @dzhou121: should we move this issue to `lapce/piet-wgpu`?)

**Comment by panekj** [comment]
Should be fixed
